### ISSUE TITLE
Immer's produce() or Lodash's cloneDeep()? 

### ISSUE THREAD

**Issue by Sleepful** [immerjs/immer#619]
## 🙋‍♂ Question

I am looking at options to have immutable state in ReactJs. One of the options is using Immer's `produce()` function and another option is using Lodash's `cloneDeep()` function. I understand that the style in which they are used is slightly difference, however I am having trouble deciding because I am struggling to understand if they differ in functionality beyond their signature.

Thank you!

**Comment by mweststrate** [comment]
cloneDeep will always fully clone your entire state. This causes two
fundamental problems which makes it quite unusable to use with react for
any slightly significant amount of data:

1. deepClone is very expensive. You have a collection of 100 items? All
those items will continuously be retreated on every update. Hitting both
the performance and the garbage collector badly
2. deepClone can't leverage memoization, because the "same" objects will
not be refererrentially equal after every update. So it means that react
will always need to rerender all your components. Which basically kills all
benefits that immutability would offer you in the first place.

In other words, deepClone, unlike produce, doesn't offer structural sharing
of all the unchanged parts of your state after an update

**Comment by Sleepful** [comment]
Thanks for your answer, so it seems it is a matter of performance mostly.

Would you consider that `produce()` is more readable than `cloneDeep()`, therefore reducing possible mistakes by developers?

I have read that `cloneDeep()` might lead to code that is hard to maintain but I do not understand why. An example is [this sarcastic post about clean code](https://javascript.info/ninja-code#reuse-names), it mentions that cloning leads to issues. If you have any insight it is welcome, perhaps to compare this with the way that `produce()` works.

**Comment by enoh-barbu** [comment]
[This benchmark](https://www.measurethat.net/Benchmarks/Show/10865/0/lodash-clonedeep-vs-immer-produce) says that `cloneDeep` is more efficient

**Comment by mweststrate** [comment]
First of all, that is comparing apples with oranges, as cloneDeep won't do structural sharing, so it renders further optimisations likes memozation downstream impossible. 

Benchmarks like that are pretty useless. Give `data1` a few hundred siblings and the picture will change pretty significantly, since `cloneDeep` is `O(total size of tree)` where `produce` is `O(depth of the change)`. So if you have a benchmark where the total size equals the actual effected nodes, then yes, cloneDeep has less overhead per node and will thus be more efficient. But that is not a meaningful scenario.

**Comment by chbdetta** [comment]
Ooopsie

https://www.measurethat.net/Benchmarks/Show/21030/0/lodash-clonedeep-vs-immer-produce-with-heavy-load

**Comment by FilipMaslovaric** [comment]
> Ooopsie
> 
> https://www.measurethat.net/Benchmarks/Show/21030/0/lodash-clonedeep-vs-immer-produce-with-heavy-load

Oopsie, woopsie!

> Benchmarks like that are pretty useless.

> Give `data1` a few hundred siblings and the picture will change pretty significantly

https://www.measurethat.net/Benchmarks/Show/23976/0/lodash-clonedeep-vs-immer-produce-with-heavy-load-sligh

| Test case name | Result                                     |
|----------------|--------------------------------------------|
| immer.produce() | immer.produce() x 291,017 ops/sec ±0.75%    |
| _.cloneDeep    | _.cloneDeep x 20.33 ops/sec ±1.58%          |

**Comment by vasilii-kovalev** [comment]
In my pet project, I have a computationally expensive function, that frequently updates a draft state.
I've created multiple versions of the function:
- `cloneDeep` from [es-toolkit](https://es-toolkit.dev) ([source](https://github.com/vasilii-kovalev/state-manager-examples/blob/20e17ee288a6c9dfd43b9cd7a086dff02ce7803b/react/src/pages/redux/utilities/move-worklogs-to-group-clone-deep.ts))
- Immer ([source](https://github.com/vasilii-kovalev/state-manager-examples/blob/20e17ee288a6c9dfd43b9cd7a086dff02ce7803b/react/src/pages/redux/utilities/move-worklogs-to-group-immer.ts))
- [Mutative](https://mutative.js.org) ([source](https://github.com/vasilii-kovalev/state-manager-examples/blob/20e17ee288a6c9dfd43b9cd7a086dff02ce7803b/react/src/pages/redux/utilities/move-worklogs-to-group-mutative.ts))
- JavaScript spread syntax ([source](https://github.com/vasilii-kovalev/state-manager-examples/blob/20e17ee288a6c9dfd43b9cd7a086dff02ce7803b/react/src/pages/redux/utilities/move-worklogs-to-group-spread.ts))

There are also a couple of other implementations ([source 1](https://github.com/vasilii-kovalev/state-manager-examples/blob/20e17ee288a6c9dfd43b9cd7a086dff02ce7803b/react/src/pages/redux/utilities/move-worklogs-to-group-actions.ts), [source 2](https://github.com/vasilii-kovalev/state-manager-examples/blob/20e17ee288a6c9dfd43b9cd7a086dff02ce7803b/react/src/pages/redux/utilities/move-worklogs-to-group-clone-deep-with-maps.ts)), but they are not 1-to-1 comparable with the listed ones.

I've created benchmarks using [Vitest](https://vitest.dev/guide/features.html#benchmarking) ([source](https://github.com/vasilii-kovalev/state-manager-examples/blob/20e17ee288a6c9dfd43b9cd7a086dff02ce7803b/react/src/pages/redux/utilities/benchmarks/move-worklogs-to-group.bench.ts)). You can find the scenario's description and the results there. Also tried [mitata](https://github.com/evanwashere/mitata), and it showed the same trend, so I decided to stick to Vitest, since I already use it.

Here are the results for only listed implementations:

```
RUN  v3.2.4

     name                                                      hz     min     max    mean     p75     p99    p995    p999     rme  samples
   · moveWorklogsToGroup (cloneDeep) [es-toolkit: 1.39.9]  6.1776  146.63  187.56  161.88  169.17  187.56  187.56  187.56  ±5.72%       10
   · moveWorklogsToGroup (Immer) [Immer: 10.1.1]           1.1135  879.50  944.39  898.09  901.71  944.39  944.39  944.39  ±1.55%       10
   · moveWorklogsToGroup (Mutative) [Mutative: 1.2.0]      1.3545  718.55  761.84  738.30  759.82  761.84  761.84  761.84  ±1.67%       10
   · moveWorklogsToGroup (spread)                          2.5012  375.87  422.74  399.81  411.63  422.74  422.74  422.74  ±2.94%       10

BENCH  Summary

  moveWorklogsToGroup (cloneDeep) [es-toolkit: 1.39.9]
    2.47x faster than moveWorklogsToGroup (spread)
    4.56x faster than moveWorklogsToGroup (Mutative) [Mutative: 1.2.0]
    5.55x faster than moveWorklogsToGroup (Immer) [Immer: 10.1.1]
```

Maybe the benchmarks in the comments to this issue are too simple, or I'm doing something wrong, but as we can see, `cloneDeep` outperforms Immer. `cloneDeep` also outperforms the implementation with spread operator, and it looks legit, since copying is more expensive than mutation, as I discovered in [my another benchmark](https://github.com/vasilii-kovalev/array-to-object-js-performance?tab=readme-ov-file#results). INP metric in Chrome DevTools (Performance tab) also confirms these trends (when measuring production builds).

If someone decides to tinker with the repository, check the README files in the [repository](https://github.com/vasilii-kovalev/state-manager-examples/blob/20e17ee288a6c9dfd43b9cd7a086dff02ce7803b/README.md) and [React folder](https://github.com/vasilii-kovalev/state-manager-examples/blob/20e17ee288a6c9dfd43b9cd7a086dff02ce7803b/react/README.md) to learn about the application and why the function has the implementation it has.

I've seen some work going on in #1152. Maybe it'll help somehow.
